tools/load_balancing_tool/optimize_model.py

```diff
@@ -0,0 +1,361 @@
+    except KeyError, e:
+        logger.critical("ERROR: layout class %s not found in %s\n",
+                        layout, layout_module)
+        raise e
+
+    solver = solverclass()
+
+    for c in solver.get_required_components():
+        assert data.has_key(c), "ERROR: component %s not found in data" % c
+
+    solver.set_data(data)
+    return solver
+
+class ModelData:
+    """
+    Convert dictionary data entry into usable object
+    """
+    def __init__(self, name, model_dict):
+        self.name = name
+        self.blocksize = model_dict['blocksize']
+        self.nthrds = model_dict['nthrds'][0]
+        ntasks = copy.deepcopy(model_dict['ntasks'])
+        cost = copy.deepcopy(model_dict['cost'])
+        assert len(ntasks) == len(cost), "ntasks data not same length as cost for %s" % name
+        # sort smallest ntasks to largest
+        tup = zip(*sorted(zip(cost, ntasks),
+                          key=operator.itemgetter(1)))
+        self.cost = list(tup[0])
+        self.ntasks = list(tup[1])
+        for j in self.ntasks:
+            if j % self.blocksize:
+                logger.warning("WARNING: %s pe %d not divisible by "
+                               "blocksize %d. Results may be invalid\n",
+                               name, j, self.blocksize)
+
+class OptimizeModel(object):
+    STATE_UNDEFINED = 0
+    STATE_UNSOLVED = 1
+    STATE_SOLVED_OK = 2
+    STATE_SOLVED_BAD = 3
+    states = ['Undefined', 'Unsolved', 'Solved', 'No Solution']
+
+    def __init__(self):
+        self.models = {}
+        self.state = self.STATE_UNDEFINED
+
+    def set_data(self, data_dict):
+        """
+        Add data to the model.
+        data_dict is dictionary of components with their data
+        example: {'totaltasks':64
+                  'ICE': {'ntasks': [2,4,8],
+                          'costs':  [10.0,6.0,4.0],
+                          'nthrds': [1,1,1],
+                          'blocksize': 8}
+                  'LND': {...}
+                 }
+
+        data is extrapolated as needed for n=1 and n=totaltasks
+        sets state to STATE_UNSOLVED
+        """
+        # get deep copy, because we need to divide ntasks by blocksize
+        self.maxtasks = data_dict['totaltasks']
+
+        for key in data_dict:
+            if isinstance(data_dict[key], dict) and 'ntasks' in data_dict[key]:
+                self.models[key] = ModelData(key, data_dict[key])
+
+
+        # extrapolate for n=1 and n=maxtasks
+        for m in self.models.values():
+            m.extrapolated = [False] * len(m.cost)
+
+            # add in data for ntasks=1 if not provided
+            if m.ntasks[0] > 1:
+                m.cost.insert(0, m.ntasks[0] * m.cost[0])
+                m.ntasks.insert(0, 1)
+                m.extrapolated.insert(0, True)
+
+            # add in data for maxtasks if not available
+            # assume same scaling factor as previous interval
+            if m.ntasks[-1] < self.maxtasks:
```

Comment:
There seems to be a bug here.  If len(m.ntasks) <= 1 then the code will crash on line 130.

Reply:
Yes, that's a bug. Thanks for the catch.

Reply:
I think the first failure with ne30_g16 left the directory in an unknown state that the solver didn't handle correctly. (the test_lbt_2 directory is there but not the timing directory). I need to fix this.